src/AppInstallerCommonCore/Certificates.cpp

```diff
@@ -0,0 +1,511 @@
+        THROW_LAST_ERROR_IF(!m_certificateContext);
+        return *this;
+    }
+
+    PinningDetails& PinningDetails::SetPinning(PinningVerificationType type)
+    {
+        m_pinning = type;
+        return *this;
+    }
+
+    bool PinningDetails::LoadFrom(const Json::Value& configuration)
+    {
+        std::string validationName = "Validation";
+
+        if (!configuration.isMember(validationName))
+        {
+            AICLI_LOG(Core, Warning, << "Details JSON item has no member " << validationName);
+            return false;
+        }
+
+        auto validationValue = JSON::GetValue<std::vector<std::string>>(configuration[validationName]);
+        if (!validationValue)
+        {
+            AICLI_LOG(Core, Warning, << "Details JSON item member " << validationName << " was not an array of strings");
+            return false;
+        }
+
+        for (const std::string& singleValidation : validationValue.value())
+        {
+            auto validationType = GetTypeFromString(singleValidation);
+
+            if (!validationType)
+            {
+                AICLI_LOG(Core, Warning, << "Details JSON validation is unknown: " << singleValidation);
+                return false;
+            }
+
+            m_pinning |= validationType.value();
+        }
+
+        if (m_pinning == PinningVerificationType::None)
+        {
+            // No need to load a certificate if not doing any pinning
+            return true;
+        }
+
+        std::string embeddedCertificateName = "EmbeddedCertificate";
+
+        if (!configuration.isMember(embeddedCertificateName))
+        {
+            AICLI_LOG(Core, Warning, << "Details JSON item has no member " << embeddedCertificateName);
+            return false;
+        }
+
+        auto embeddedCertificateValue = JSON::GetValue<std::string>(configuration[embeddedCertificateName]);
+        if (!validationValue)
+        {
+            AICLI_LOG(Core, Warning, << "Details JSON item member " << embeddedCertificateName << " was not a string");
+            return false;
+        }
+
+        auto embeddedCertificateBytes = Utility::ParseFromHexString(embeddedCertificateValue.value());
+        LoadCertificate(embeddedCertificateBytes);
+
+        return true;
+    }
+
+    bool PinningDetails::Validate(PCCERT_CONTEXT certContext) const
+    {
+        if (WI_IsFlagSet(m_pinning, PinningVerificationType::PublicKey))
+        {
+            if (!CertComparePublicKeyInfo(
+                X509_ASN_ENCODING | PKCS_7_ASN_ENCODING,
+                &m_certificateContext.get()->pCertInfo->SubjectPublicKeyInfo,
+                &certContext->pCertInfo->SubjectPublicKeyInfo))
+            {
+                AICLI_LOG(Core, Verbose, << "Public key mismatch: Expected certificate [" << GetSimpleDisplayName(m_certificateContext.get()) << "], Actual certificate [" << GetSimpleDisplayName(certContext) << "]");

```

Comment:
>Verbose [](http://example.com/codeflow?start=32&length=7)

Should these logging levels at least warning?

Reply:
I chose not to log them by default for the case of multiple chains.  It is expected that only one of the chains would be valid for any given connection, so logging "Hey there is seemingly a security issue over here!" is not warranted when it turns out that it is actually just the one not being used.

`PinningConfiguration::Validate`, which is the expected entry point for any non-test use case, does log at `Error` on a failure to find any good chain.  If that is happening, then the user can manually gather a verbose log.